### DISCUSSION TITLE
Error converting DB Json to object "Could not convert document"

### DISCUSSION THREAD

**Discussion by mfonnesbeck** [ravendb/ravendb#16853] [Q&A]
We are migrating from 3.X to 5.X and have encountered a problem deserializing our database object from JSON to object and cannot find any help online to figure out the conversion.  **The error we are receiving is:**

Could not convert document 28a9d267-2d8a-4379-bdb5-260488463c56 to entity of type EmpoweredHoa.Web.Core.Models.InvoiceApproval. Json: {"TicketId":null,"ApprovedTimestamp":null,"ApprovedByUserId":null,"Votes":{"2023-07-12T16:17:33.7675357Z":[]},"@metadata":{"@collection":"InvoiceApprovals","Raven-Clr-Type":"EmpoweredHoa.Web.Core.Models.InvoiceApproval, EmpoweredHoa.Web.Core","@change-vector":"A:6205-LxE+fvEr+EirswfyKiiWGw","@id":"28a9d267-2d8a-4379-bdb5-260488463c56","@last-modified":"2023-07-12T16:17:37.2927156Z"}}
at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.SessionBlittableJsonConverter.FromBlittable(Type type, BlittableJsonReaderObject& json, String id, Boolean trackEntity)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\SessionBlittableJsonConverter.cs:line 75
at Raven.Client.Documents.Session.InMemoryDocumentSessionOperations.TrackEntity(Type entityType, String id, BlittableJsonReaderObject document, BlittableJsonReaderObject metadata, Boolean noTracking)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Documents\Session\InMemoryDocumentSessionOperations.cs:line 532
at Raven.Client.Documents.Session.InMemoryDocumentSessionOperations.TrackEntity(Type entityType, DocumentInfo documentFound)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Documents\Session\InMemoryDocumentSessionOperations.cs:line 484
at Raven.Client.Documents.Session.InMemoryDocumentSessionOperations.TrackEntity[T](DocumentInfo documentFound)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Documents\Session\InMemoryDocumentSessionOperations.cs:line 446
at Raven.Client.Documents.Session.Operations.LoadOperation.GetDocument[T](String id)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Documents\Session\Operations\LoadOperation.cs:line 146
at Raven.Client.Documents.Session.Operations.LoadOperation.GetDocuments[T]()
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Documents\Session\Operations\LoadOperation.cs:line 183
at Raven.Client.Documents.Session.DocumentSession.Load[T](IEnumerable`1 ids)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Documents\Session\DocumentSession.Load.cs:line 54

**With an inner exception of:**
Could not find the right converter from cache for 'System.Collections.Generic.IReadOnlyDictionary`2[[System.DateTime, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.Collections.Generic.List`1[[EmpoweredHoa.Web.Core.Models.Vote, EmpoweredHoa.Web.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]], mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]'.

at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.Converters.CachingJsonConverter.ThrowConverterIsNullException(Type objectType)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\Converters\CachingJsonConverter.cs:line 89
at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.Converters.CachingJsonConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\Converters\CachingJsonConverter.cs:line 53
at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.DeserializeConvertable(JsonConverter converter, JsonReader reader, Type objectType, Object existingValue)
at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.NewtonsoftJsonJsonSerializer.Raven.Client.Json.Serialization.IJsonSerializer.Deserialize(IJsonReader reader, Type type)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\NewtonsoftJsonJsonSerializer.cs:line 10
at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.NewtonsoftJsonBlittableEntitySerializer.EntityFromJsonStream(Type type, BlittableJsonReaderObject json)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\NewtonsoftJsonBlittableEntitySerializer.cs:line 64
at Raven.Client.Json.Serialization.NewtonsoftJson.NewtonsoftJsonSerializationConventions.Raven.Client.Json.Serialization.ISerializationConventions.DeserializeEntityFromBlittable(Type type, BlittableJsonReaderObject json)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Json\Serialization\NewtonsoftJson\NewtonsoftJsonSerializationConventions.cs:line 155
at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.SessionBlittableJsonConverter.FromBlittable(Type type, BlittableJsonReaderObject& json, String id, Boolean trackEntity)
      in C:\Builds\RavenDB-Stable-5.4\54049\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\SessionBlittableJsonConverter.cs:line 56

The object in question has a dictionary of a DateTime and a list of Votes, which is another object that has a string for the UserId, DateTime timestamp, and a Vote Value which is an enumeration. But as you can see we have an object with a timestamp of when the Vote started and no Votes entered yet.

The class has a `private Dictionary<DateTime, List<Vote>> votes = new Dictionary<DateTime, List<Vote>>();`
And then has a getter and setter to hide the setting:
```
        public IReadOnlyDictionary<DateTime, List<Vote>> Votes
        {
            get { return votes; }
            private set { votes = new Dictionary<DateTime, List<Vote>>((Dictionary<DateTime, List<Vote>>)value); }
        }
```

It appears this is where it is having issues converting from the JSON to the object. Attempting to look through documentation and finding not really anything of help I could use some direction or advice.

Thank you.

**Comment by ayende** [comment]
Try customizing the serializer with:

```
  documentStore.Conventions.Serialization = new NewtonsoftJsonSerializationConventions
                {
                    CustomizeJsonSerializer = serializer => serializer.ObjectCreationHandling=ObjectCreationHandling.Reuse
                };

```

**Reply by mfonnesbeck** [reply]
We actually already have setup code that adds custom converters for serialization and deserialization due to some complications of a couple of objects. What does setting the Reuse option do?  Can we do that in addition to adding to the serializers' Converters collection?

**Reply by ayende** [reply]
Can you show a failing test with this issue?
I double checked the code, the error location doesn't make sense.

**Reply by mfonnesbeck** [reply]
[InvoiceApproval.txt](https://github.com/ravendb/ravendb/files/12031378/InvoiceApproval.txt)
Using this definition of the Invoice Approval object, and the Json that Raven is reporting its trying to load:
{"TicketId":null,"ApprovedTimestamp":null,"ApprovedByUserId":null,"Votes":{"2023-07-12T16:17:33.7675357Z":[]},"@metadata":{"@collection":"InvoiceApprovals","Raven-Clr-Type":"EmpoweredHoa.Web.Core.Models.InvoiceApproval, EmpoweredHoa.Web.Core","@change-vector":"A:6205-LxE+fvEr+EirswfyKiiWGw","@id":"28a9d267-2d8a-4379-bdb5-260488463c56","@last-modified":"2023-07-12T16:17:37.2927156Z"}}

It fails with that error. Using RavenDB Client 5.4.104.

**Reply by garayx** [reply]
Please submit a unit test or sample application that demonstrate the issue.
Here you can find a guide on how to create unit test using TestDriver:
https://ravendb.net/docs/article-page/5.4/Csharp/start/test-driver#complete-example

**Reply by mfonnesbeck** [reply]
I cannot. I created a test using the Raven Test Driver and instantiated an Invoice Approval object just like the JSON that is being returned, saved it to the database and then attempted to load it using the Guid Id, and it loaded just fine. I also created a test that saves several and then selects out all their Ids and loads the Invoice Approval objects using those Linq-Selected Ids and it also worked fine. So, now I'm at a loss of what is different and wrong with our environment or database.  Our database was one created at ravendb.cloud and then imported in a dump file from our 3.X development database. I don't know if that creates any complications. But that is what we are working with.

**Reply by ayende** [reply]
The reason we ask for the test it to be able to see what is going on.
Given your scenario, you are doing something in the initialization of your system that cause this problem.
The test shows that this isn't a stock RavenDB issue, but something specific to your configuration. Keep adding your configuration to the test, until you find what item make it fail

**Reply by mfonnesbeck** [reply]
Ok. I was able to reproduce the issue with a Unit Test. I found that I was using the same document session to save and load the document, and it worked. When I used one document session to save and then another to load the document by Id, it failed with the error. Even with an embedded Raven database it failed. I am now trying to simplify my Unit Test project so that I can upload it here. 

I am curious about two things. 
1) In our dependency injection configuration when we are setting up our database store. We set the Conventions with a serialization and include a Custom Serializer for a StringEnumConverter for handling storing a particular enumeration for a certain document type and then a CustomizeJsonDeserializer to our own Converter for Security Claims that are store in User documents. Would setting up these converters possibly be causing any problems?

2) I am curious about the inner exception that I posted above for this deserialization of this object: "Could not find the right converter from cache". That seems to suggest that we might need a converter for this object, since it seems to have a problem with the deserialization of the Dictionary<DateTime, List<Vote>>. Thoughts?

**Reply by mfonnesbeck** [reply]
Ok. Here's a test project that produces the issue.  Let me know what you think.  This is the last problem holding us up with our migration. Thank you.
[InvoiceUnitTests.zip](https://github.com/ravendb/ravendb/files/12053172/InvoiceUnitTests.zip)

**Reply by ayende** [reply]
Okay, that is a really interesting problem.
The basic issue is that you have a type mismatch deep inside JSON.Net

At one point, we are being asked about the type using `IReadOnlyDictionary<DateTime, List<Vote>>`
The next, we are asked using the concrete value `Dictionary<DateTime, List<Vote>>`
Finally, we are being asked on `IReadOnlyDictionary<DateTime, List<Vote>>`

The problem is that you have different types and it is not expecting this.

I changed your code to be:

```
    public class InvoiceApproval
    {
        private Dictionary<DateTime, List<Vote>> votes;

        public IReadOnlyDictionary<DateTime, List<Vote>> Votes
        {
            get { return votes; }
            private set { votes = new Dictionary<DateTime, List<Vote>>(value); }
        }
```

Alternatively, you can register an explicit converter for `IReadOnlyDictionary`

Thanks for the test was quite helpful

**Reply by mfonnesbeck** [reply]
When I make those changes above by removing the class initialization of the votes member and remove the casting of the value in the set method. I get a compiler error stating, "cannot convert from 'System.Collections.Generic.IReadOnlyDictionary<System.DateTime, System.Collections.Generic.List<EmpoweredHoa.Web.Core.Models.Vote>>' to 'int' "  

This doesn't seem to be a fix for my issue. Am I missing something?

**Reply by ayende** [reply]
Huh? Where are you getting the erro?

**Reply by mfonnesbeck** [reply]
It was a compiler error. Last night I was thinking about this and I think it was a problem with the .Net framework version we had for the Unit Test project. I wasn't able to create the same version as our main code base. I figured out this morning that the newer .Net version of the test project allows for the creation of a Dictionary<DateTime, List<Vote>> from an object of that type or similar without type casting. It was the type casting error that I was getting in my main code base. I was able to figure things out from there and we got it working.  Thank you.
